nexus/blueprint-execution/src/resource_allocation.rs

```diff
@@ -0,0 +1,998 @@
+        // allocated, then both attempt to create it? We believe this is
+        // okay: the loser of the race (i.e., the one whose create tries to
+        // commit second) will fail to allocate the IP, which will bubble
+        // out and prevent realization of the current blueprint. That's
+        // exactly what we want if two Nexuses try to realize the same
+        // blueprint at the same time.
+        if self
+            .is_external_ip_already_allocated(
+                zone_type,
+                service_id,
+                external_ip,
+                None,
+            )
+            .await?
+        {
+            return Ok(());
+        }
+        let ip_id = Uuid::new_v4();
+        let description = zone_type;
+        self.datastore
+            .allocate_explicit_service_ip(
+                self.opctx,
+                ip_id,
+                ip_name,
+                description,
+                service_id,
+                external_ip,
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to allocate IP to {zone_type} {service_id}: \
+                         {external_ip}"
+                )
+            })?;
+
+        info!(
+            self.log, "successfully allocated external IP";
+            "zone_type" => zone_type,
+            "zone_id" => %service_id,
+            "ip" => %external_ip,
+            "ip_id" => %ip_id,
+        );
+
+        Ok(())
+    }
+
+    // BoundaryNtp uses a SNAT service IPs; this method is similar to
+    // `ensure_external_service_ip` but accounts for that.
+    async fn ensure_external_service_snat_ip(
+        &self,
+        zone_type: &'static str,
+        service_id: Uuid,
+        snat: &SourceNatConfig,
+    ) -> anyhow::Result<()> {
+        // Only attempt to allocate `external_ip` if it isn't already assigned
+        // to this zone.
+        //
+        // This is subject to the same kind of TOCTOU race as described for IP
+        // allocation in `ensure_external_service_ip`, and we believe it's okay
+        // for the same reasons as described there.
+        if self
+            .is_external_ip_already_allocated(
+                zone_type,
+                service_id,
+                snat.ip,
+                Some((snat.first_port, snat.last_port)),
+            )
+            .await?
+        {
+            return Ok(());
+        }
+
+        let ip_id = Uuid::new_v4();
+        self.datastore
+            .allocate_explicit_service_snat_ip(
+                self.opctx,
+                ip_id,
+                service_id,
+                snat.ip,
+                (snat.first_port, snat.last_port),
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to allocate snat IP to {zone_type} {service_id}: \
+                     {snat:?}"
+                )
+            })?;
```

Comment:
Same comment applies here I guess, and maybe it does suggest that there is (or should be) a higher-level note.